src/cpp/src/speculative_decoding/speculative_decoding_stateful_eagle3.cpp

```diff
@@ -0,0 +1,1042 @@
+}
+
+InferResult Eagle3TargetWrapper::forward(const InferContext& ctx) {
+    eagle3::log_debug(eagle3::PipelineStep::ITER,
+                      "Target forward: input_token_count=" + std::to_string(ctx.input_token_count) + ", sample_count=" +
+                          std::to_string(ctx.sample_count) + ", validate=" + std::to_string(ctx.num_tokens_to_validate),
+                      m_verbose);
+
+    // 1. Prepare inputs from sequence state
+    ov::Tensor input_ids, attention_mask, position_ids;
+    build_model_inputs(ctx.input_token_count, input_ids, attention_mask, position_ids);
+
+    // 2. Infer
+    auto output = infer(input_ids, attention_mask, position_ids);
+
+    // 3. Sample (use sample_count for number of positions to sample from)
+    auto sampled = sample_tokens(output.logits, ctx.input_token_count, ctx.sample_count, ctx.num_tokens_to_validate);
+
+    // 4. Store hidden states to sequence for draft model to use
+    get_current_sequence()->update_hidden_state(output.hidden_features);
+
+    return InferResult{std::move(output), std::move(sampled)};
+}
+
+Eagle3DraftWrapper::Eagle3DraftWrapper(const ov::genai::ModelDesc& model_desc) : Eagle3InferWrapperBase(model_desc) {
+    eagle3::log_debug(eagle3::PipelineStep::INIT, "Draft model wrapper initialized", m_verbose);
+}
+
+void Eagle3DraftWrapper::initialize_sequence(const ov::Tensor& input_ids, const ov::genai::GenerationConfig& config) {
+    const auto shape = input_ids.get_shape();
+    OPENVINO_ASSERT(shape.size() == 2 && shape[0] == 1,
+                    "Expected input_ids shape [1, seq_len], got ",
+                    eagle3::format_shape(shape));
+
+    const int64_t* ids_data = input_ids.data<const int64_t>();
+    const size_t total_len = shape[1];
+    OPENVINO_ASSERT(total_len >= 2, "Draft model requires at least 2 tokens");
+
+    // Draft model uses tokens[1:] (Eagle3 specific behavior)
+    TokenIds draft_prompt_ids(ids_data + 1, ids_data + total_len);
+    m_sequence_group = std::make_shared<SequenceGroup>(1, draft_prompt_ids, config, 0);
+
+    // TODO(top-k): For top-k, draft model may need to track multiple candidate sequences
+    OPENVINO_ASSERT(get_running_sequence_count() == 1,
+                    "Expected single sequence after initialization, got ",
+                    get_running_sequence_count());
+
+    eagle3::log_debug(eagle3::PipelineStep::INIT,
+                      "Draft sequence initialized: prompt_len=" + std::to_string(draft_prompt_ids.size()) +
+                          " (from original " + std::to_string(total_len) + ")",
+                      m_verbose);
+}
+
+InferenceOutput Eagle3DraftWrapper::infer(const ov::Tensor& input_ids,
+                                          const ov::Tensor& attention_mask,
+                                          const ov::Tensor& position_ids,
+                                          const ov::Tensor& hidden_states) {
+    const size_t input_token_count = input_ids.get_shape()[1];
+
```

Comment:
is it possible that draft wrapper and target wrapper share some common operations? like standard inputs? common forward etc